File: oxidize-golang/core/model/inference_config.go
package model

import (
	"math"
	"strings"

	"github.com/Zapdev-labs/oxidize/golang/core/ggufcore"
)

// applyTokenEmbeddingDims fills hidden/vocab from token_embd when metadata is missing.
// GGUF stores token_embd as [embedding_length, vocab_size] (see oxidize-core inference.rs).
func applyTokenEmbeddingDims(out *InferenceConfig, dims []uint64) {
	if len(dims) < 2 {
		return
	}
	if out.HiddenSize == 0 {
		out.HiddenSize = int(dims[0])
	}
	if out.VocabSize == 0 {
		out.VocabSize = int(dims[1])
	}
}

// InferenceConfigFromGGUF builds config from GGUF metadata and tensor shapes.
func InferenceConfigFromGGUF(mapped *ggufcore.MappedFile) InferenceConfig {
	file := mapped.Parsed
	out := DefaultInferenceConfig()
	arch := ggufcore.Architecture(file)
	if arch == "" {
		arch = "llama"
	}
	out.Architecture = architectureFromGGUFString(arch)

	archKey := func(suffix string) string { return arch + "." + suffix }
	archU32 := func(suffix string) (uint32, bool) {
		if v, ok := file.Metadata[archKey(suffix)]; ok {
			if n, ok := v.AsUint64(); ok {
				return uint32(n), true
			}
		}
		if v, ok := file.Metadata["llama."+suffix]; ok {
			if n, ok := v.AsUint64(); ok {
				return uint32(n), true
			}
		}
		return 0, false
	}
	archF32 := func(suffix string) (float32, bool) {
		if v, ok := file.Metadata[archKey(suffix)]; ok {
			if f, ok := v.AsFloat32(); ok {
				return f, true
			}
		}
		if v, ok := file.Metadata["llama."+suffix]; ok {
			if f, ok := v.AsFloat32(); ok {
				return f, true
			}
		}
		return 0, false
	}

	if n, ok := archU32("context_length"); ok {
		out.ContextSize = int(n)
	}
	if n, ok := archU32("embedding_length"); ok {
		out.HiddenSize = int(n)
	}
	if n, ok := archU32("attention.head_count"); ok {
		out.NumAttentionHeads = int(n)
	}
	if n, ok := archU32("attention.head_count_kv"); ok {
		out.NumKeyValueHeads = int(n)
	} else {
		out.NumKeyValueHeads = out.NumAttentionHeads
	}
	if n, ok := archU32("attention.key_length"); ok {
		out.KeyValueHeadDim = int(n)
	}
	if n, ok := archU32("feed_forward_length"); ok {
		out.IntermediateSize = int(n)
	}
	if n, ok := archU32("block_count"); ok {
		out.LayerCount = int(n)
	}
	if f, ok := archF32("attention.layer_norm_rms_epsilon"); ok {
		out.RMSNormEps = f
	}
	if f, ok := archF32("rope.freq_base"); ok {
		out.RopeTheta = f
	}
	if n, ok := archU32("vocab_size"); ok {
		out.VocabSize = int(n)
	}
	if n, ok := archU32("attention.sliding_window"); ok {
		out.SlidingWindow = int(n)
	}
	if n, ok := archU32("expert_count"); ok {
		out.NumExperts = int(n)
	}
	if n, ok := archU32("expert_used_count"); ok {
		out.NumExpertsPerToken = int(n)
	}
	if out.NumExpertsPerToken == 0 {
		out.NumExpertsPerToken = 2
	}
	if out.NumExperts == 0 {
		for _, info := range ggufcore.MappedTensorInfos(file) {
			if strings.HasPrefix(info.Name, "blk.0.") && strings.Contains(info.Name, "ffn_gate_exps") {
				if n, ok := archU32("expert_count"); ok {
					out.NumExperts = int(n)
				}
				break
			}
		}
	}

	for _, info := range ggufcore.MappedTensorInfos(file) {
		switch info.Name {
		case "token_embd.weight", "tok_embeddings.weight", "model.embed_tokens.weight":
			applyTokenEmbeddingDims(&out, info.Dimensions)
		}
	}
	if out.VocabSize == 0 {
		out.VocabSize = 32000
	}

	// Gemma 2/3/4 specifics: interleaved local/global attention with dual RoPE
	// theta, embedding scaling, GeGLU, and sandwich normalization. Gated behind
	// the Gemma architecture so other models are unaffected.
	if out.Architecture == ArchGemmaModel {
		pattern := 6
		if arch == "gemma2" {
			pattern = 2
		}
		if n, ok := archU32("attention.sliding_window_pattern"); ok && n > 0 {
			pattern = int(n)
		}
		out.SlidingWindowPattern = pattern
		if f, ok := archF32("rope.freq_base_swa"); ok && f > 0 {
			out.RopeThetaSWA = f
		} else {
			out.RopeThetaSWA = 10000.0
		}
		if out.HiddenSize > 0 {
			out.EmbeddingScale = float32(math.Sqrt(float64(out.HiddenSize)))
		}
		out.GeluFFN = true
		out.SandwichNorm = true
	}
	return out
}

func architectureFromGGUFString(arch string) Architecture {
	switch arch {
	case "llama":
		return ArchLlamaModel
	case "mistral":
		return ArchMistralModel
	case "mixtral":
		return ArchMixtralModel
	case "deepseek", "deepseek_v2", "deepseek_v3", "deepseek_moe":
		return ArchDeepSeekModel
	case "qwen", "qwen2", "qwen2moe", "qwen3", "qwen35":
		return ArchQwenModel
	case "gemma", "gemma2", "gemma3", "gemma4":
		return ArchGemmaModel
	case "phi", "phi2", "phi3":
		return ArchPhiModel
	case "falcon":
		return ArchFalconModel
	case "gpt2":
		return ArchGpt2Model
	case "gptj":
		return ArchGptJModel
	case "gptneox":
		return ArchGptNeoXModel
	case "minimax", "minimax-m2", "minimax-text-01":
		return ArchMiniMaxModel
	default:
		return ArchLlamaModel
	}
}

// LoadInferenceFromGGUF loads weights and returns a ready InferenceModel.
func LoadInferenceFromGGUF(mapped *ggufcore.MappedFile) (*InferenceModel, error) {
	config := InferenceConfigFromGGUF(mapped)
	decoderCfg := LlamaDecoderConfigFromInference(config)
	stack, err := LoadLlamaDecoderStackFromGGUF(mapped, decoderCfg)
	if err != nil {
		return nil, err
	}
	if stack.Output.IsLoaded() && config.VocabSize == 0 {
		config.VocabSize = stack.Output.outputDim()
	}
	if stack.TokEmbeddings.IsLoaded() && config.VocabSize == 0 {
		config.VocabSize = stack.TokEmbeddings.outputDim()
	}
	return NewInferenceModel(config, WeightStorage{File: mapped}, stack), nil
}

File: oxidize-python/oxidize_python/core/model/inference_config.py
"""GGUF-derived inference config mirroring oxidize-golang inference_config.go."""

from __future__ import annotations

import math
from dataclasses import dataclass
from enum import StrEnum

from oxidize_python.core.ggufcore import gguf as ggufcore
from oxidize_python.core.ggufcore.gguf import MappedFile, mapped_tensor_infos
from oxidize_python.core.kv_cache import Quantization as KvQuantization
from oxidize_python.core.model.model import Architecture
from oxidize_python.internal.gguf.types import MetadataValue


class KvCacheDType(StrEnum):
    F32 = "f32"
    F16 = "f16"
    Q8 = "q8"
    Q4 = "q4"


@dataclass
class InferenceConfig:
    vocab_size: int = 32000
    context_size: int = 2048
    layer_count: int = 32
    hidden_size: int = 4096
    intermediate_size: int = 11008
    num_attention_heads: int = 32
    num_key_value_heads: int = 32
    key_value_head_dim: int = 128
    kv_cache_dtype: str = "f32"
    kv_quantization: KvQuantization = KvQuantization.ASYMMETRIC
    rms_norm_eps: float = 1e-5
    rope_theta: float = 10000.0
    architecture: Architecture = Architecture.LLAMA
    sliding_window: int = 0
    num_experts: int = 0
    num_experts_per_token: int = 1
    alibi_num_heads: int = 0
    # Gemma-family fields (see oxidize-core inference.rs).
    rope_theta_swa: float = 0.0
    sliding_window_pattern: int = 0
    embedding_scale: float = 1.0
    gelu_ffn: bool = False
    sandwich_norm: bool = False

    def head_dim(self) -> int:
        if self.num_attention_heads == 0:
            return 0
        return self.hidden_size // self.num_attention_heads

    def kv_head_dim(self) -> int:
        if self.key_value_head_dim != 0:
            return self.key_value_head_dim
        return self.head_dim()


def default_inference_config() -> InferenceConfig:
    return InferenceConfig()


def _meta_uint64(v: MetadataValue | None) -> int | None:
    if v is None:
        return None
    n, ok = v.as_uint64()
    return int(n) if ok else None


def _meta_float32(v: MetadataValue | None) -> float | None:
    if v is None:
        return None
    f, ok = v.as_float32()
    return f if ok else None


def architecture_from_gguf_string(arch: str) -> Architecture:
    match arch:
        case "llama":
            return Architecture.LLAMA
        case "mistral":
            return Architecture.MISTRAL
        case "mixtral":
            return Architecture.MIXTRAL
        case "deepseek" | "deepseek_v2" | "deepseek_v3" | "deepseek_moe":
            return Architecture.DEEPSEEK
        case "qwen" | "qwen2" | "qwen2moe" | "qwen3" | "qwen35":
            return Architecture.QWEN
        case "gemma" | "gemma2" | "gemma3" | "gemma4":
            return Architecture.GEMMA
        case "phi" | "phi2" | "phi3":
            return Architecture.PHI
        case "falcon":
            return Architecture.FALCON
        case "gpt2":
            return Architecture.GPT2
        case "gptj":
            return Architecture.GPTJ
        case "gptneox":
            return Architecture.GPT_NEOX
        case "minimax" | "minimax-m2" | "minimax-text-01":
            return Architecture.MINIMAX
        case _:
            return Architecture.LLAMA


def _apply_token_embedding_dims(out: InferenceConfig, dims: list[int]) -> None:
    """GGUF token_embd dims are [embedding_length, vocab_size] (oxidize-core)."""
    if len(dims) < 2:
        return
    if out.hidden_size == 0:
        out.hidden_size = int(dims[0])
    if out.vocab_size == 0:
        out.vocab_size = int(dims[1])


def inference_config_from_gguf(mapped: MappedFile) -> InferenceConfig:
    file = mapped.parsed
    out = default_inference_config()
    arch = ggufcore.architecture(file) or "llama"
    out.architecture = architecture_from_gguf_string(arch)

    def arch_key(suffix: str) -> str:
        return f"{arch}.{suffix}"

    def arch_u32(suffix: str) -> int | None:
        for key in (arch_key(suffix), f"llama.{suffix}"):
            if n := _meta_uint64(file.metadata.get(key)):
                return n
        return None

    def arch_f32(suffix: str) -> float | None:
        for key in (arch_key(suffix), f"llama.{suffix}"):
            if f := _meta_float32(file.metadata.get(key)):
                return f
        return None

    if n := arch_u32("context_length"):
        out.context_size = n
    if n := arch_u32("embedding_length"):
        out.hidden_size = n
    if n := arch_u32("attention.head_count"):
        out.num_attention_heads = n
    if n := arch_u32("attention.head_count_kv"):
        out.num_key_value_heads = n
    else:
        out.num_key_value_heads = out.num_attention_heads
    if n := arch_u32("attention.key_length"):
        out.key_value_head_dim = n
    if n := arch_u32("feed_forward_length"):
        out.intermediate_size = n
    if n := arch_u32("block_count"):
        out.layer_count = n
    if f := arch_f32("attention.layer_norm_rms_epsilon"):
        out.rms_norm_eps = f
    if f := arch_f32("rope.freq_base"):
        out.rope_theta = f
    if n := arch_u32("vocab_size"):
        out.vocab_size = n
    if n := arch_u32("attention.sliding_window"):
        out.sliding_window = n
    if n := arch_u32("expert_count"):
        out.num_experts = n
    if n := arch_u32("expert_used_count"):
        out.num_experts_per_token = n

    for info in mapped_tensor_infos(file):
        if info.name in (
            "token_embd.weight",
            "tok_embeddings.weight",
            "model.embed_tokens.weight",
        ):
            _apply_token_embedding_dims(out, info.dimensions)
    if out.vocab_size == 0:
        out.vocab_size = 32000

    # Gemma 2/3/4: interleaved local/global attention with dual RoPE theta,
    # embedding scaling, GeGLU, and sandwich normalization. Gated behind the
    # Gemma architecture so other models are unaffected.
    if out.architecture == Architecture.GEMMA:
        pattern = 2 if arch == "gemma2" else 6
        if n := arch_u32("attention.sliding_window_pattern"):
            pattern = n
        out.sliding_window_pattern = pattern
        if f := arch_f32("rope.freq_base_swa"):
            out.rope_theta_swa = f
        else:
            out.rope_theta_swa = 10000.0
        if out.hidden_size > 0:
            out.embedding_scale = math.sqrt(out.hidden_size)
        out.gelu_ffn = True
        out.sandwich_norm = True
    return out


def load_inference_from_gguf(mapped: MappedFile):
    from oxidize_python.core.model.gguf_decoder_load import load_llama_decoder_stack_from_gguf
    from oxidize_python.core.model.inference import InferenceModel, WeightStorage

    config = inference_config_from_gguf(mapped)
    decoder_cfg = llama_decoder_config_from_inference(config)
    stack, err = load_llama_decoder_stack_from_gguf(mapped, decoder_cfg)
    if err:
        raise err
    if stack.output.is_loaded() and config.vocab_size == 0:
        config.vocab_size = stack.output._output_dim()
    if stack.tok_embeddings.is_loaded() and config.vocab_size == 0:
        config.vocab_size = stack.tok_embeddings._output_dim()
    return InferenceModel(config, WeightStorage(file=mapped), stack)


def llama_decoder_config_from_inference(cfg: InferenceConfig):
    from oxidize_python.core.model.llama_decoder import LlamaDecoderConfig

    return LlamaDecoderConfig(
        hidden_size=cfg.hidden_size,
        layer_count=cfg.layer_count,
        intermediate_size=cfg.intermediate_size,
        num_attention_heads=cfg.num_attention_heads,
        num_key_value_heads=cfg.num_key_value_heads,
        key_value_head_dim=cfg.kv_head_dim(),
        vocab_size=cfg.vocab_size,
        rms_norm_eps=cfg.rms_norm_eps,
        rope_theta=cfg.rope_theta,
        sliding_window=cfg.sliding_window,
        rope_theta_swa=cfg.rope_theta_swa,
        sliding_window_pattern=cfg.sliding_window_pattern,
        embedding_scale=cfg.embedding_scale,
        gelu_ffn=cfg.gelu_ffn,
        sandwich_norm=cfg.sandwich_norm,
    )
